📝 Update docs/parser/supported-formats/rails.mdx (#554)
liam-hq/liam: frontend/apps/docs/content/docs/parser/supported-formats/rails.mdx

title: schema.rb (Ruby on Rails)
---

If you’re using Ruby on Rails with ActiveRecord, in most cases you can automatically generate a useful ER diagram with Liam ERD. This page provides instructions and tips for generating an ER diagram in a Rails project.

## ActiveRecord and db/schema.rb

When a Rails application manages migrations using ActiveRecord, the latest database structure is typically described in `db/schema.rb`.  
Since `db/schema.rb` is recommended to be kept under version control, most projects will have it in their Git repository.

When using Liam CLI, specify `--format schemarb` and `--input db/schema.rb` as follows:

<Tabs items={["npm", "pnpm", "yarn", "bun"]}>
  <Tab value="npm" copyable>npx @liam-hq/cli erd build --format schemarb --input db/schema.rb</Tab>
  <Tab value="pnpm" copyable>pnpm dlx @liam-hq/cli erd build --format schemarb --input db/schema.rb</Tab>
  <Tab value="yarn" copyable>yarn dlx @liam-hq/cli erd build --format schemarb --input db/schema.rb</Tab>
  <Tab value="bun" copyable>bun x @liam-hq/cli erd build --format schemarb --input db/schema.rb</Tab>

> </Tabs>

Author: no syntax color. but it seems to be ok. https://github.com/liam-hq/liam/pull/539#issuecomment-2609102959

If the above command runs without issue, you should see an ER diagram generated.

## Under the Hood

- Liam CLI does not run an internal Ruby or Rails runtime. In other words, there is no Ruby process running under the hood.
- Instead, it analyzes the content of `db/schema.rb` using the [`ruby/prisma` parser](https://github.com/ruby/prism).

> ## If You Don’t Have `db/schema.rb` but Use `db/structure.sql`

Reviewer: This one looks good for troubleshooting too.

```suggestion
## Trouble Shooting: If You Don’t Have `db/schema.rb` but Use `db/structure.sql`
```

If your Rails app doesn’t have `db/schema.rb` and instead uses `db/structure.sql`, here’s what you need to know.

### For Rails 7.0 or Later

Somewhere under `config` (most commonly `config/application.rb`), you may have `config.active_record.schema_format = :sql`.  
In such cases, running the dump command `rails db:schema:dump` will generate `db/structure.sql`.

If you want to obtain `db/schema.rb`, run the following in an environment with database access:

- Change `config.active_record.schema_format` to `:ruby` and run `rails db:schema:dump`.
- Alternatively, set the environment variable `SCHEMA_FORMAT=ruby` and run `rails db:schema:dump`.

>   - For example: `SCHEMA_FORMAT=ruby rails db:schema:dump`

Author: I haven't actually created a sample project to verify it yet, but it should be correct, so I'd like to merge it into the main branch. I'll check the correctness later today.

  <Tab value="pnpm" copyable>pnpm dlx @liam-hq/cli erd build --format postgres --input db/structure.sql</Tab>
  <Tab value="yarn" copyable>yarn dlx @liam-hq/cli erd build --format postgres --input db/structure.sql</Tab>
  <Tab value="bun" copyable>bun x @liam-hq/cli erd build --format postgres --input db/structure.sql</Tab>
</Tabs>

## Pro Tips: Using `Schemafile` Instead of `db/schema.rb`

<Callout title="warn" type="warn">Please note that this approach can be somewhat complicated or may not work as expected.</Callout>

If your project uses a tool called [Ridgepole](https://github.com/ridgepole/ridgepole), you might have a `Schemafile` instead of a `db/schema.rb`.

A `Schemafile` uses a DSL similar to what appears in `db/schema.rb`, so handing `Schemafile` directly to Liam CLI might work in some cases.

However, because users can write arbitrary Ruby code in a `Schemafile`, there may be scenarios where it isn’t fully compatible. If the output is not what you expect, consider generating a `db/schema.rb` using the methods described above (e.g., `rails db:schema:dump`) and then parsing that file.

> ## Pro Tips: When Associations Don’t Appear in the ER Diagram

Reviewer: Since users who want this information are more troubled than professional, how about the following?

```suggestion
## Trouble Shooting: When Associations Don’t Appear in the ER Diagram
```

```ruby
# Prerequisite: The database must be accessible.

Rails.application.eager_load!

# Filter out models that have valid table_name
models_with_table = ActiveRecord::Base.descendants.select do |model|
  model.table_exists? && model.table_name.present?
end

# Map belongs_to associations to the corresponding referenced tables
assoc_map = models_with_table.map do |model|
  target_tables = model.reflect_on_all_associations(:belongs_to).map do |assoc|
    assoc.class_name.safe_constantize&.table_name
  end.compact

  [model.table_name, target_tables]
end.to_h

# Generate add_foreign_key statements from the mapped relationships
# This is a simple example, so in practice you’d need to handle duplicates carefully
content = assoc_map.flat_map do |from_table, to_tables|
  to_tables.map { |to_table| %(add_foreign_key "#{from_table}", "#{to_table}") }
end.join("\n")

# Write or append this content to db/schema.rb in some way
```

By taking these steps to define foreign keys, you can produce an ER diagram closer to your expectations.

> ---
> 
> Try the above tips and tricks to get the most out of Liam ERD in your Rails projects. If foreign keys are clearly defined, you can generate a concise, easy-to-understand visualization of your database structure.

Reviewer: I think we can delete this one.

```suggestion
```

Author: 👍
